File: handlers/payment.py
from aiogram import types, Router, Dispatcher
from aiogram.filters import Command
from database.database import cursor, connection

router = Router()

@router.message(Command(commands=['pay']))
async def handle_pay(message: types.Message):
    parts = message.text.split()
    
    if len(parts) < 3:
        await message.answer("❌ Incorrect command format. Example: /pay 2.5 (coins) jnRHDtjlCfhNfc5b (Pay id)")
        return

    try:
        coin_amount = float(parts[1])
        user_key = parts[2]
    except ValueError:
        await message.answer("❌ Specify the correct number for coin.")
        return

    if coin_amount < 0.01:
        await message.answer("❌ Minimum transfer amount: 0.01 coin.")
        return

    sender_id = message.from_user.id

    cursor.execute("SELECT coins FROM Users_coins WHERE id_users = ?", (sender_id,))
    sender_data = cursor.fetchone()

    if sender_data is None or sender_data[0] < coin_amount:
        await message.answer("❌ There are not enough coins to transfer.")
        return

    new_sender_balance = sender_data[0] - coin_amount
    cursor.execute("UPDATE Users_coins SET coins = ? WHERE id_users = ?", (round(new_sender_balance, 2), sender_id))

    cursor.execute("SELECT coins FROM Users_coins WHERE id_pay = ?", (user_key,))
    receiver_data = cursor.fetchone()

    if receiver_data:
        new_receiver_balance = receiver_data[0] + coin_amount
        cursor.execute("UPDATE Users_coins SET coins = ? WHERE id_pay = ?", (round(new_receiver_balance, 2), user_key))
    else:
        await message.answer("❌ User with this Pay ID does not exist.")
        connection.commit()
        return

    connection.commit()

    await message.answer(
        f"✅ Successfully transferred {coin_amount} tokens to the user with the ID PAY {user_key}.\n"
        f"💳 Your new balance: {round(new_sender_balance, 2)} coins."
    )

def register_handlers(dp: Dispatcher):
    dp.include_router(router)

File: src/handlers/payment.rs
use std::f32::NAN;

use frankenstein::{Api, Message, SendMessageParams, TelegramApi};

pub async fn handle_pay(api: &Api, message: Message) {
    let text = match message.text.clone() {
        Some(t) => t,
        None => return,
    };

    let parts: Vec<&str> = text.split_whitespace().collect();

    if parts.len() < 3 {
        let params = SendMessageParams {
            chat_id: frankenstein::ChatId::Integer(message.chat.id),
            text: "❌ Incorrect command format. Example: /pay 2 (coins) jnRHDtjlCfhNfc5b (Pay id)"
                .to_string(),
            parse_mode: None,
            entities: None,
            disable_web_page_preview: None,
            disable_notification: None,
            reply_to_message_id: None,
            allow_sending_without_reply: None,
            reply_markup: None,
        };
        let _ = api.send_message(&params);
        return;
    }

    let coin_amount = match parts[1].parse::<i64>() {
        Ok(val) => val,
        Err(_) => {
            let params = SendMessageParams {
                chat_id: frankenstein::ChatId::Integer(message.chat.id),
                text: "❌ Specify the correct number for coin.".to_string(),
                parse_mode: None,
                entities: None,
                disable_web_page_preview: None,
                disable_notification: None,
                reply_to_message_id: None,
                allow_sending_without_reply: None,
                reply_markup: None,
            };
            let _ = api.send_message(&params);
            return;
        }
    };

    if coin_amount == 0 {
        let params = SendMessageParams {
            chat_id: frankenstein::ChatId::Integer(message.chat.id),
            text: "❌ Minimum transfer amount: 1 coin.".to_string(),
            parse_mode: None,
            entities: None,
            disable_web_page_preview: None,
            disable_notification: None,
            reply_to_message_id: None,
            allow_sending_without_reply: None,
            reply_markup: None,
        };
        let _ = api.send_message(&params);
        return;
    }

    let user_key = parts[2];
    let sender_id = message.from.as_ref().map(|u| u.id).unwrap_or(0);

    let sender_balance = crate::db::get_user_balance(sender_id as i64).unwrap_or(0);
    if sender_balance < coin_amount {
        let params = SendMessageParams {
            chat_id: frankenstein::ChatId::Integer(message.chat.id),
            text: "❌ There are not enough coins to transfer.".to_string(),
            parse_mode: None,
            entities: None,
            disable_web_page_preview: None,
            disable_notification: None,
            reply_to_message_id: None,
            allow_sending_without_reply: None,
            reply_markup: None,
        };
        let _ = api.send_message(&params);
        return;
    }
    if crate::db::get_user_idpay(sender_id as i64)
        .unwrap()
        .unwrap()
        == user_key
    {
        let params = SendMessageParams {
            chat_id: frankenstein::ChatId::Integer(message.chat.id),
            text: "❌ Can't send coins to yourself".to_string(),
            parse_mode: None,
            entities: None,
            disable_web_page_preview: None,
            disable_notification: None,
            reply_to_message_id: None,
            allow_sending_without_reply: None,
            reply_markup: None,
        };
        let _ = api.send_message(&params);
        return;
    }

    let (success, new_sender_balance) =
        crate::db::transfer_coins(sender_id as i64, user_key, coin_amount).unwrap_or((false, None));
    if !success {
        let params = SendMessageParams {
            chat_id: frankenstein::ChatId::Integer(message.chat.id),
            text,
            parse_mode: None,
            entities: None,
            disable_web_page_preview: None,
            disable_notification: None,
            reply_to_message_id: None,
            allow_sending_without_reply: None,
            reply_markup: None,
        };
        let _ = api.send_message(&params);
        return;
    }

    let response = format!(
        "✅ Successfully transferred {} tokens to the user with the ID PAY {}.\n💳 Your new balance: {} coins.",
        coin_amount,
        user_key,
        new_sender_balance.unwrap_or(sender_balance)
    );
    let params = SendMessageParams {
        chat_id: frankenstein::ChatId::Integer(message.chat.id),
        text: response,
        parse_mode: None,
        entities: None,
        disable_web_page_preview: None,
        disable_notification: None,
        reply_to_message_id: None,
        allow_sending_without_reply: None,
        reply_markup: None,
    };

    let _ = api.send_message(&params);
}
